### PR TITLE
feat: add cloud recommendation message to CI output

### LINKED ISSUE #24000: Add cloud recommendation message to CI output
[ClickUp](https://app.clickup.com/t/18033298/PM-2552)

Add the following message to the CI output for `cypress run`:
```bash
-----------------------------------------------------------------------------------------

Having trouble debugging your CI failures?
Record your runs to the Cypress Cloud to watch video recordings for each test, debug failing and flaky tests, and integrate with your favorite tools.

>> https://on.cypress.io/cloud-get-started

------------------------------------------------------------------------------------------
```
This message should show after the test summary (in place of the recorded run link) in run mode if the following conditions are met:
- we are in run mode
- the run is **not** being recorded to the dashboard
- we are running in CI
- at least one test failed during the run
- `CYPRESS_NO_COMMERCIAL_RECOMMENDATIONS` environment variable is not set to `true` (users can opt out of seeing this message by setting this true)

### PR DESCRIPTION
**Pull request by astone123**
- Closes https://github.com/cypress-io/cypress/issues/24000

### User facing changelog

n/a

### Additional details

This PR adds a message that prompts the user to go to Cypress Cloud after the test summary table only when **all** of the following conditions are true:
- Cypress is in run mode
- The process is running in CI (using `is-ci` library to determine this)
- `CYPRESS_COMMERCIAL_RECOMMENDATIONS` environment variable is **not** set to `0`
- At least one test in the run failed
- `--record` flag was not passed to the cypress run command (the user is not attempting to record this run to Cypress Cloud)

Services PR for on link: https://github.com/cypress-io/cypress-services/pull/4972
Docs PR: https://github.com/cypress-io/cypress-documentation/pull/4863

### Steps to test

Look at the output from [this run](https://app.circleci.com/pipelines/github/cypress-io/cypress/45747/workflows/fc2c3edc-752b-4adc-887a-0fc9cc3d31e4/jobs/1924189) and verify that the cloud recommendation message appears after the test summary. Note that Circle CI is redacting `cypress.io` from the on link. This is a setting in our Circle CI project and shouldn't be an issue in other's projects.

You can also look at the tests that I added in `cypress_spec.js` and run those

### How has the user experience changed?

Screen Shot 2022-11-17 at 2 46 33 PM

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4863
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/server/lib/util/print-run.ts
Comment (lmiller1990) [inline packages/server/lib/util/print-run.ts]: cc @pstakoun, good to get product :eyes:  for any public facing marketing call to action messages.

File: packages/server/lib/util/print-run.ts
Comment (lmiller1990) [inline packages/server/lib/util/print-run.ts]: Hmm what do you think either

1. make `record` non nullable
2. or have a default value

Just a personal preference, but I like for booleans to be explicit, so we can communicate the expected default value to future developers.

File: packages/server/lib/util/print-run.ts
Comment (lmiller1990) [inline packages/server/lib/util/print-run.ts]: Shouldn't this be `!record`? Eg, if we are NOT recording, return. Otherwise, we might log the message.

File: packages/server/lib/util/print-run.ts
Comment (lmiller1990) [inline packages/server/lib/util/print-run.ts]: Nit, but same here - can we make this non optional or have a default value?

File: packages/server/test/integration/cypress_spec.js
Comment (lmiller1990) [inline packages/server/test/integration/cypress_spec.js:397]: Nice job finding a way to test this, unit testing `run.js` is often tricky.

**Review by lmiller1990 — CHANGES_REQUESTED** [review]
I *think* there is logic bug - please confirm I'm not going crazy. 

Other than that, code and tests look :100:

File: packages/server/lib/util/print-run.ts
Comment (astone123) [reply packages/server/lib/util/print-run.ts]: No, if the run is being recorded (if `--recorded` is passed, making `record = true`) then we want to exit early and **not** show the message.

We only want to suggest setting up on Cypress Cloud if they are **not already** recording the run. Maybe I could refactor the logic here to make things more clear?

File: packages/server/lib/util/print-run.ts
Comment (ZachJW34) [inline packages/server/lib/util/print-run.ts]: Nit: I feel like this function call doesn't belong in this function but rather right after the `renderSummaryTable`.

File: packages/server/test/integration/cypress_spec.js
Comment (ZachJW34) [inline packages/server/test/integration/cypress_spec.js]: Does `process.env` get restored after each test? When manipulating global state, I like to have an `after` or `afterEach` that sets the global state back to what it should be such that, if the test fails, it doesn't pollute any other state

**Review by ZachJW34 — APPROVED** [review]
Tested it out, works great! Couple small comments

**Comment by ZachJW34** [comment]
@astone123 Does deleting the CircleCI `CYPRESS_NO_COMMERCIAL_RECOMMENDATIONS` have any impact on this being merged?

**Comment by astone123** [comment]
@ZachJW34 yeah, I need that variable in the project in order for the tests to pass, so I need to put it back before I push a follow-up commit, but that will cause everyone else's tests to fail until this branch gets merged... I'll put something in Slack to explain the situation once this is ready to merge

**Comment by astone123** [comment]
Waiting on this for now - going to meet with @pstakoun and Brian tomorrow to get final approval

File: packages/server/test/integration/cypress_spec.js
Comment (astone123) [reply packages/server/test/integration/cypress_spec.js]: It looks like `process.env` is restored after each test. For instance, if I have `CYPRESS_NO_COMMERCIAL_RECOMMENDATIONS=true` set in the Circle project environment and then I have
```js
delete process.env.CYPRESS_NO_COMMERCIAL_RECOMMENDATIONS
```
in a test, when the next test runs, `process.env` will still have
```js
{
  CYPRESS_NO_COMMERCIAL_RECOMMENDATIONS: true
}
```
even though I deleted it in the test before it

File: packages/server/test/integration/cypress_spec.js
Comment (astone123) [reply packages/server/test/integration/cypress_spec.js:397]: yeah, it was tricky. one thing that I found (and I still don't understand why this happens) is that I couldn't get my test case to work like this

```js
it('does something', () => {
  // my test
})
```
and I had to declare the callback function like this

```js
it('does something', function () {
  // my test
})
```

it took me a while to figure out that my test was failing because I was trying to define the callback as an arrow function 🤔 still don't know why

File: packages/server/lib/util/print-run.ts
Comment (astone123) [reply packages/server/lib/util/print-run.ts]: Good call 👍🏻

File: packages/server/lib/util/print-run.ts
Comment (lmiller1990) [reply packages/server/lib/util/print-run.ts]: Right, gotcha, figured I was misreading this (I guess it's not immediately clear, or I just derped out... seems fine).

**Review by lmiller1990 — APPROVED** [review]
💯

**Review by brian-mann — COMMENTED** [review]
.

**Review by brian-mann — CHANGES_REQUESTED** [review]
- Might want to switch to using the `is-ci` npm utility since we use that elsewhere
- Consider where to document this env var in the docs (maybe installation, with another section in Configuration that enumerates env vars that control behavior - with links to other areas)
- The name of the env var sounds good - we may want to align this to the existing `CYPRESS_CRASH_REPORTS` since that doesn't use the "NO" verbiage
- Instead of hard coding the line dividers, go through the same utility methods bc I believe they dynamically adjust their width to the available space in the terminal
- Let's iterate on the colors, formatting, line dividers, to make the message less aggressive
- After iterating just drop in a picture of the formatting and we can do async reviews

**Comment by astone123** [comment]
@brian-mann thoughts on this appearance?

Screen Shot 2022-11-16 at 4 07 28 PM

File: tooling/v8-snapshot/src/setup/force-no-rewrite.ts
Comment (ryanthemanuel) [inline tooling/v8-snapshot/src/setup/force-no-rewrite.ts:72]: Why do these need to be in the force-no-rewrite category?

Comment (astone123) [reply tooling/v8-snapshot/src/setup/force-no-rewrite.ts:72]: I'm trying to figure out exactly why, but it wouldn't work without adding those (they probably don't _all_ need to be in there).

I think what might be happening is that the exported value from `is-ci` is getting evaluated and snapshotted before `process.env.CI=true` is set, and then it never gets re-evaluated? 

If I don't have these in this file then `is-ci` returns `false` even though we are running in CI and it should be `true`. @ZachJW34 originally suggested trying this when we were working through it

Comment (ZachJW34) [reply tooling/v8-snapshot/src/setup/force-no-rewrite.ts:72]: @astone123 and I paired on this a bit. Not sure we need every one of these but it wasn't working without this addition. I believe it's due to how `is-ci` and `ci-info` are written, where the export is evaluated upon require e.g. `module.exports = !!process.env.CI` and not a function.

Could be totally wrong though!

Comment (ryanthemanuel) [reply tooling/v8-snapshot/src/setup/force-no-rewrite.ts:72]: I think from playing around locally all we need to add here is the first two and I'm good with that for now. Though I still want to figure a little more on why.

**Review by brian-mann — APPROVED** [review]

**Pull request merged**
